Compfiles/Usa2018P1.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2024 The Compfiles Contributors. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hongyu Ouyang
+-/
+
+import Mathlib.Tactic
+import Mathlib.Analysis.SpecialFunctions.Pow.Real
+
+import ProblemExtraction
+
+problem_file { tags := [.Algebra, .Inequality] }
+
+/-!
+# USA Mathematical Olympiad 2018, Problem 1
+
+Given that a,b,c are positive real numbers such that
+
+  a + b + c = 4 ∛(abc)
+
+prove that 2(ab + bc + ca) + 4min(a²,b²,c²) ≥ a² + b² + c²
+-/
+
+namespace Usa2018P1
+
+snip begin
+
+lemma am_gm (a b : ℝ) : a > 0 → b > 0 → 2 * (a * b) ^ ((1 : ℝ) / 2) ≤ a + b := by
+  intro ha hb
+  have gsa := Real.sqrt_pos.mpr ha
+  have gsb := Real.sqrt_pos.mpr hb
+  set sa := Real.sqrt a with hsa
+  set sb := Real.sqrt b with hsb
+  have : a = sa * sa := by rw [hsa]; field_simp
+  rw [this]
+  have : b = sb * sb := by rw [hsb]; field_simp
+  rw [this]
+  have : (sa * sa * (sb * sb)) ^ ((1 : ℝ) / 2) = sa * sb := by
+    simp only [one_div]
+    rw [Real.mul_rpow, ←pow_two, ←pow_two, ←Real.rpow_two, ←Real.rpow_two]
+    rw [Real.rpow_rpow_inv (by positivity), Real.rpow_rpow_inv (by positivity)]
+    linarith
+    linarith
+    positivity
+    positivity
+  rw [this]
+  rw [←pow_two, ←pow_two]
+  rw [←mul_assoc]
+  refine two_mul_le_add_sq sa sb
+
+snip end
+
+
+problem usa2018_p1 (a b c : ℝ) : a > 0 → b > 0 → c > 0 → a + b + c = 4 * (a * b * c) ^ ((1 : ℝ) / 3) →
+  2 * (a * b + b * c + c * a) + 4 * (min (min (a * a) (b * b)) (c * c)) ≥ a^2 + b^2 + c^2 := by
+  -- solution 1 from
+  -- https://artofproblemsolving.com/wiki/index.php/2018_USAMO_Problems/Problem_1
+  intro ha hb hc heq
+  wlog h1 : a ≤ b with H1
+  · have swap1 : (a * b + b * c + c * a) = (b * a + a * c + c * b) := by ring 
+    have swap2 : (min (min (a * a) (b * b)) (c * c)) = (min (min (b * b) (a * a)) (c * c)) := by
```

Comment:
A possibly nicer way to say `(min (min (a * a) (b * b)) (c * c))` is `sInf {a*a, b*b, c*c}`.